beacon_node/eth1/src/service.rs

```diff
@@ -990,34 +999,41 @@ impl Service {
             .filter(|x| *x <= latest_in_cache)
             .take(max_blocks_per_update)
             .collect::<Vec<_>>();
+
+        debug!(
+            self.log,
+            "Downloading eth1 blocks";
+            "first" => ?required_block_numbers.first(),
+            "last" => ?required_block_numbers.last(),
+        );
+
         // Produce a stream from the list of required block numbers and return a future that
         // consumes the it.
 
-        let eth1_blocks: Vec<Eth1Block> = stream::try_unfold(
-            required_block_numbers.into_iter(),
-            |mut block_numbers| async {
-                match block_numbers.next() {
-                    Some(block_number) => {
-                        match endpoints
-                            .first_success(|e| async move {
-                                download_eth1_block(e, self.inner.clone(), Some(block_number)).await
-                            })
-                            .await
-                        {
-                            Ok(eth1_block) => Ok(Some((eth1_block, block_numbers))),
-                            Err(e) => Err(e),
-                        }
-                    }
-                    None => Ok(None),
-                }
-            },
-        )
-        .try_collect()
-        .await
-        .map_err(Error::FallbackError)?;
-
         let mut blocks_imported = 0;
-        for eth1_block in eth1_blocks {
+        let start_instant = Instant::now();
+        // Setting the maximum runtime longer than the auto update interval allows us to download
+        // more blocks without interruption.
+        //
+        // This value is a trade-off between making blocks download quickly and starving the deposit
+        // cache of updates.
+        let max_runtime = Duration::from_millis(self.config().auto_update_interval_millis * 2);
+        for block_number in required_block_numbers {
+            // Avoid updates that lasts longer than the update interval.
+            //
+            // This prevents the block downloading routine for running for a very long time and
+            // starving the deposit cache updater.
+            if Instant::now().duration_since(start_instant) > max_runtime {
```

Comment:
I'm not sure I understand why we can't complete the blocks downloads before going back to deposits.

Even if we go back to updating deposits every 14secs, we usually won't get more than 1-2 new logs every update and this would slow down the block download.

Reply:
> I'm not sure I understand why we can't complete the blocks downloads before going back to deposits.

I suspect we could do this and I had a look at it, but it was sufficiently complicated that I wasn't confident to do it before mainnet :)

Reply:
I meant we could just remove the 
`if Instant::now().duration_since(start_instant) > max_runtime` 
check and let all the block downloads complete. If we get an error somewhere in between, we start from that point in the next update.

Reply:
I had some reasoning behind this, but it was loosely held. I think I'll just follow your suggestion and leave it out :)